### ISSUE TITLE
AtomicAdjust -> std::atomic?

### ISSUE THREAD

**Issue by CFSworks** [panda3d/panda3d#258]
Since Panda3D is now a C++11 project, would it make sense to deprecate AtomicAdjust, replacing all of the various uses with `std::atomic`? This would allow greater portability to platforms that aren't GCC/x86/POSIX and hence don't have the necessary intrinsics, would allow Panda to benefit from performance improvements in the compiler, and of course is less code for us to maintain and less Panda-specific functionality for a new developer to learn.

As an intermediate step, I could pretty easily check in an "AtomicAdjustCxx11Impl" where `Integer` is just a typedef on `std::atomic_int` and the static methods are just wrappers around the `std::atomic`'s operations. Does this sound like a good idea, or is there actually some reason we want to keep the AtomicAdjust family around?

**Comment by theclashingfritz** [comment]
Along with that the 2010 compiler minimum should be bumped to 2013 for full C++11 feature support.

**Comment by rdb** [comment]
I'm in favour of this in principle.  Part of the benefit of moving to C++11 is being able to ditch a lot of threading code in Panda.

Using C++11 atomics requires dropping support for MSVC 2010 and for macOS Snow Leopard.  I'm fine with that in principle, but we should open issues about that to see whether everyone is OK with that.  (MSVC doesn't fully support C++11 concurrency features until 2015, and I'm personally fine with setting the minimum to 2015, which still supports targeting XP.)

I tried making an AtomicAdjustCxx11Impl before as you describe, but it's not that simple since there is code that makes the assumption that `Integer`/`Pointer` is a plain type, using it to store intermediates, so that code would need to be reworked too.  If we want to do something without removing backward compat, we could introduce a `patomic` type, which is typedeffed to `std::atomic` on compilers that support it or implemented using `AtomicAdjust` otherwise.

**Comment by CFSworks** [comment]
I'm not sure what you mean by "intermediates" - do you mean there are stack variables that are of `AtomicAdjust::Integer` type? This sounds like it might be a good thing to refactor nonetheless since, although it is a plain type, it's also got alignment attributes to ensure lockless access, and I worry that the extra stack-juggling the compiler goes through to ensure that alignment might be an unnecessary slowdown in some cases.

I doubt anybody would complain about the minimum compiler version becoming MSVC 2015. I'm not as sure about dropping macOS <10.7. I really doubt anybody's on that these days, but since we're worried about losing XP it'd be inconsistent of us not to worry about old, old versions of macOS.

I don't have very strong feelings one way or another about `patomic`, but I think doing that would constitute commitment toward getting rid of `AtomicAdjust` eventually (which I'm good with as long as we can move forward to full C++11 atomics). But if we're worried about long-term backwards compatibility, it'd make sense to implement an `AtomicAdjust` atop `std::atomic` just to keep the implementation-switching easy to understand (it'd be confusing to have one file decide between atomic/"AtomicAdjustWrapper", then the wrapper that wraps AtomicAdjust, then another file that decides what variety of AtomicAdjust implementation to use).

**Comment by rdb** [comment]
Yes, I'm referring to temporaries and function arguments that use `AtomicAdjust::{Integer,Pointer}`, probably so that they can make sure to have the same size.  There's a fair amount of casting to that type as well, and some non-atomic loads/stores of those variables.  Alignment is not really a performance concern here since all platforms we currently build for align loads/stores of word-size integers to the word size boundary anyway.

I was suggesting `patomic` precisely because it constitutes said commitment to get rid of AtomicAdjust and lets us gradually move the code to the C++11-style syntax, but you're right that it is confusing to have both syntaxes around.  I don't feel strongly one way or the other on this, but either way we'll have to do some work to refactor code to be able to use a C++11-based AtomicAdjust implementation.

Let's open separate issues on dropping Snow Leopard support and on raising the minimum MSVC version, so that people can weigh in on that.  If it turns out we can drop those, then perhaps we can just switch everything over to the C++11 syntax in one fell swoop.

**Comment by rdb** [comment]
The point that hasn't been raised in this discussion yet that using AtomicAdjust lets us turn off atomic access when compiling Panda without threading.  Switching to `std::atomic` means that applications incur the overhead of atomic operations even when compiling Panda without threading.

Use of a patomic wrapper could solve this.

**Comment by rdb** [comment]
AtomicAdjust is gone, replaced with patomic instead, which is a typedef for `std::atomic` in the threaded build.
